### PR TITLE
Initial work on a libjpeg-turbo fuzzer.

### PR DESCRIPTION
**Pull request by alex**
Unfortunately I have not yet succeeded in testing this locally; running ` python infra/helper.py build_fuzzers libjpeg-turbo` locally seems to go ok, but does not produce anything in `build/out/`. Would appreciate some guidance on the ways in which I'm too stupid to fuzz :-)

### CODE REVIEW

**Comment by kcc** [comment]
>> not yet succeeded in testing this locally
Did you try just running it w/o docker manually, as described in libfuzzer.info? 

Does this change brings in a seed corpus?

Is there a chance to have the fuzz target (the .cc file) upstream? 
Frankly, I am reluctant to bring any more .cc targets -- we need to push to have all of them upstream.

**Comment by alex** [comment]
The following works fine:

```
$ python infra/helper.py shell libjpeg-turbo
# compile
# /out/libjpeg_turbo_fuzzer
```

I don't know whether upstream will accept this; I'm not associated with them, this just seemed like a straightforward place to start contributing.

**Comment by kcc** [comment]
Thanks for doing this, and I think this is fine to have the .cc fuzz target for jpeg-turbo here for now
(Mike will review the rest).
But in near future we really should ban new .cc fuzz targets in oss-fuzz because we need to encourage
the upstream to own them.

File: README.md [hunk 1]
```diff
@@ -3,8 +3,8 @@
 > *Status*: Beta. We are preparing the project for the first public release. Documentation and smoothing the process is our main priority.
 
 [FAQ](docs/faq.md)
-| [New Target Guide](docs/new_target.md) 
-| [Reproducing](docs/reproducing.md) 
+| [New Target Guide](docs/new_target.md)
```
Comment (mikea) [inline README.md]: Let's separate documentation changes from libjpeg work.

File: targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#include <stdint.h>
```
Comment (mikea) [inline targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc]: This need copyright headers.

File: targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    int width, height, subsamp, colorspace;
+    int res = tjDecompressHeader3(
+        jpegDecompressor, data, size, &width, &height, &subsamp, &colorspace);
+
+    if (res != 0 || width == 0 || height == 0) {
+        tjDestroy(jpegDecompressor);
+        return 0;
+    }
+
+    // TODO: this can't possibly be right?
```
Comment (mikea) [inline targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc]: Not sure what you mean here.

File: targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    tjhandle jpegDecompressor = tjInitDecompress();
+
+    int width, height, subsamp, colorspace;
+    int res = tjDecompressHeader3(
+        jpegDecompressor, data, size, &width, &height, &subsamp, &colorspace);
+
+    if (res != 0 || width == 0 || height == 0) {
```
Comment (mikea) [inline targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc]: We will need some protection against zip-bombs. E.g. check for width * height < 1M

File: targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        tjDestroy(jpegDecompressor);
+        return 0;
+    }
+
+    // TODO: this can't possibly be right?
+    void *buf = malloc(width * height * 3);
```
Comment (mikea) [inline targets/libjpeg-turbo/libjpeg_turbo_fuzzer.cc]: since this is .cc file, new & delete please. The cast below won't be necessary too.

**Comment by alex** [comment]
All resolved!

**Comment by mikea** [comment]
Alex, thanks for working on this. I'll merge this in a minute, but keep in mind that unless you are libjpeg-turbo developer we won't be able to CC you on issues because of possible security concerns.

To take credit for your work I advise you to reach out to libjpeg-turbo mailing lists, tell them what you did and try to find engineers to assign the issues (if any).

**Comment by alex** [comment]
👍  makes sense to me. Thanks for merging!

**Pull request merged**